spec/models/work_packages/scopes/allowed_to_spec.rb

```diff
@@ -0,0 +1,138 @@
+  let(:public_non_module_action) { :view_project }
+  let(:non_module_action) { :edit_project }
+
+  context 'when querying for a permission that does not exist' do
+    it 'raises an error' do
+      expect do
+        described_class.allowed_to(build(:user), :non_existing_permission)
+      end.to raise_error(Authorization::UnknownPermissionError)
+    end
+  end
+
+  context 'when the user is an admin' do
+    let(:user) { create(:admin) }
+
+    subject { described_class.allowed_to(user, action) }
+
+    it 'returns all work packages' do
+      expect(subject).to contain_exactly(
+        work_package_in_public_project,
+        work_package_in_private_project,
+        other_work_package_in_private_project
+      )
+    end
+  end
+
+  context 'when the user has the permission directly on the work package' do
+    let(:work_package_permissions) { [action] }
+
+    before do
+      create(:member, project: private_project, entity: work_package_in_private_project,
+                      user:, roles: [work_package_role])
+    end
+
+    subject { described_class.allowed_to(user, action) }
+
+    it 'returns the authorized work package' do
+      expect(subject).to contain_exactly(work_package_in_private_project)
+    end
+  end
+
+  context 'when the user has the permission on the project the work package belongs to' do
+    let(:project_permissions) { [action] }
+
+    before do
+      create(:member, project: private_project, user:, roles: [project_role])
+    end
+
+    subject { described_class.allowed_to(user, action) }
+
+    it 'returns the authorized work packages' do
+      expect(subject).to contain_exactly(work_package_in_private_project, other_work_package_in_private_project)
+    end
+  end
+
+  context 'when the user has a different permission on the project, but the requested one on a specific work package' do
+    let(:project_permissions) { [:view_work_packages] }
+    let(:work_package_permissions) { %i[view_work_packages edit_work_packages] }
+
+    before do
+      create(:member, project: private_project, entity: work_package_in_private_project, user:, roles: [work_package_role])
+      create(:member, project: private_project, user:, roles: [project_role])
+    end
+
+    context 'and requesting a permission that is only granted on the single work package' do
+      subject { described_class.allowed_to(user, :edit_work_packages) }
+
+      it 'returns the authorized work packages' do
+        expect(subject).to contain_exactly(work_package_in_private_project)
+      end
+    end
+
+    context 'and requesting a permission that is granted on the project and the work package' do
+      subject { described_class.allowed_to(user, :view_work_packages) }
+
+      it 'returns the authorized work packages' do
+        expect(subject).to contain_exactly(work_package_in_private_project, other_work_package_in_private_project)
+      end
+    end
+  end
+
+  # TODO: Add more tests that check anonymous and non-member permissions
```

Comment:
What you said 😄 Maybe the brunt of the tests for when there is a permission on a project membership can be copied over from the specs on `Project.allowed_to`.